### PR TITLE
feat: conditionally animate discord assets

### PR DESCRIPTION
**Pull request by skdhg**
This idea is inspired by Discord itself. When the window is focused, we render animated avatar, otherwise we render static avatar. Since the contributing guide is missing, I may have done something inappropriately. Please feel free to correct me if that's the case.

~~I haven't applied this to Guild icons, let me know if I should do that in this PR.~~
As per the conversation in Discord, I have replaced `<Image/>` with `<DiscordImage/>` for user avatars/guild icons.

### CODE REVIEW

File: src/components/DiscordImage.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"use client";
+
+import Image from "next/image";
+import useFocus from "~/hooks/use-focus";
+
+type DiscordImageProps = React.ComponentProps<typeof Image>;
```
Comment (florian-lefebvre) [inline src/components/DiscordImage.tsx:6]: In the future, we'll see if we want to restrict props

File: src/components/DiscordImage.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    // ignore placeholders
+    !props.src.includes("/embed") &&
+    props.src.includes(".gif")
+      ? props.src.replace(
+          focused ? ".webp" : ".gif",
+          focused ? ".gif" : ".webp"
+        )
+      : props.src;
+
+  // eslint-disable-next-line
```
Comment (florian-lefebvre) [inline src/components/DiscordImage.tsx:22]: Why is this required? What's the error message?

File: src/components/DiscordImage.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+"use client";
+
+import Image from "next/image";
+import useFocus from "~/hooks/use-focus";
+
+type DiscordImageProps = React.ComponentProps<typeof Image>;
+
+export default function DiscordImage(props: DiscordImageProps) {
+  const focused = useFocus();
```
Comment (florian-lefebvre) [inline src/components/DiscordImage.tsx:9]:
```suggestion
  const { src, ...rest } = props;
  const focused = useFocus();
```

File: src/components/DiscordImage.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  const imageSrc =
+    typeof props.src === "string" &&
+    // ignore placeholders
+    !props.src.includes("/embed") &&
+    props.src.includes(".gif")
+      ? props.src.replace(
+          focused ? ".webp" : ".gif",
+          focused ? ".gif" : ".webp"
+        )
+      : props.src;
```
Comment (florian-lefebvre) [inline src/components/DiscordImage.tsx:20]:
```suggestion
  const imageSrc =
    typeof src === "string" &&
    // ignore placeholders
    !src.includes("/embed") &&
    src.includes(".gif")
      ? src.replace(
          focused ? ".webp" : ".gif",
          focused ? ".gif" : ".webp"
        )
      : src;
```
Won't be preperly formatted but not a problem, I'll handle that

File: src/components/DiscordImage.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  // eslint-disable-next-line
+  return <Image {...props} src={imageSrc} />;
```
Comment (florian-lefebvre) [inline src/components/DiscordImage.tsx:23]:
```suggestion
  return <Image {...rest} src={imageSrc} />;
```

File: src/hooks/use-focus.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+"use client";
+
+import { useEffect, useState } from "react";
+
+type OnFocusChangeHandler = (focused: boolean) => unknown;
```
Comment (florian-lefebvre) [inline src/hooks/use-focus.ts:5]:
```suggestion
```

File: src/hooks/use-focus.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+export default function useFocus<E extends HTMLElement>(
+  target?: React.RefObject<E> | null,
+  onFocusChange?: OnFocusChangeHandler
```
Comment (florian-lefebvre) [inline src/hooks/use-focus.ts:9]:
```suggestion
  onFocusChange?: (focused: boolean) => void;
```

File: src/hooks/use-focus.ts
Comment (florian-lefebvre) [inline src/hooks/use-focus.ts:1]: Did you take this file anywhere? If yes, would be awesome to add a comment to the top of the file with the reference link

File: src/hooks/use-focus.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+) {
+  const [focused, setFocused] = useState(() => {
+    return window.document.activeElement != null;
```
Comment (florian-lefebvre) [inline src/hooks/use-focus.ts:12]: Does this work even in SSR? I think we could face an error with window being undefined but tell me if you encountered any error

**Review by florian-lefebvre — CHANGES_REQUESTED** [review]
Awesome contribution! I added a few comments and reviews, good work!

File: src/hooks/use-focus.ts (hunk 8 above)
Comment (skdhg) [reply src/hooks/use-focus.ts:11]: Haven't had issues during testing

Comment (florian-lefebvre) [reply src/hooks/use-focus.ts:11]:
```suggestion
    return window?.document.activeElement != null;
```
Just in case

File: src/components/DiscordImage.tsx (hunk 2 above)
Comment (skdhg) [reply src/components/DiscordImage.tsx:23]: It's the `alt` prop that was giving a warning.

File: src/hooks/use-focus.ts (hunk 8 above)
Comment (florian-lefebvre) [reply src/hooks/use-focus.ts:11]: yeah you're right
```suggestion
    return typeof window !== undefined ? window.document.activeElement != null: true;
```

File: src/components/DiscordImage.tsx (hunk 2 above)
Comment (florian-lefebvre) [reply src/components/DiscordImage.tsx:23]:
```suggestion
  // TODO: check why alt is not detected
  // eslint-disable-next-line
```

**Review by florian-lefebvre — CHANGES_REQUESTED** [review]

**Review by florian-lefebvre — APPROVED** [review]
LGTM
